sympy/matrices/normalforms.py

```diff
@@ -0,0 +1,115 @@
+from sympy.matrices import *
+
+'''
+Functions returning normal forms of matrices
+
+'''
+def smith_normal_form(m, domain = None):
```

Comment:
There should also be some reference though there are not many apart from Wikipedia. That will contain further references to the theoretical background (Lang's Algebra, 3rd ed.) and applications to modules over PIDs (invariant factors, elementary divisors).